### PR TITLE
MAINT: Use AxisError in more places

### PR DESCRIPTION
**Pull request by eric-wieser**
Behavioural changes:

* Lots of code throws `AxisError` instead of `IndexError` or `ValueError` - but this is backwards-compatible, since `AxisError` derives from both (#8584)
* `_ureduce` detects duplicate arguments correctly - previously, `np.median([1, 2], axis=(0,0))` would fail, but `np.median([1, 2], axis=(-1,-1))` and `np.median([1, 2], axis=(0,-1))` would be fine. Now all three raise `ValueError`.
* `np.gradient` now accepts `axis=[1,2]` as well as `axis=(1,2)`. This might not be desirable, but is consistent with the majority of other functions. Now that this all goes through one place, we could deprecate it in future.
* `np.core.numeric._validate_axis` is slightly more public at `np.core.numeric.normalize_axis_tuple`, to complement `np.core.multiarray.normalize_axis_index`
* Functions taking a single axis argument now all call `operator.index`, rather than just a random subset of them.

### CODE REVIEW

File: numpy/core/src/multiarray/common.h [hunk 1]
```diff
@@ -158,9 +162,21 @@ check_and_adjust_axis(int *axis, int ndim)
...
-                     "axis %d is out of bounds for array of dimension %d",
-                     *axis, ndim);
+        /* build arguments for the AxisError constructor */
+        args = Py_BuildValue("iiO", *axis, ndim, msg_prefix);
+        if(args == NULL) {
+            return -1;
+        }
+
+        /* Invoke the AxisError constructor */
+        exc = PyObject_CallObject(AxisError_cls, args);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/common.h]: Implementing the string formatting in C seemed like way too much effort for little gain - no need for performance once something starts going wrong

File: numpy/lib/tests/test_function_base.py [hunk 2]
```diff
@@ -2974,11 +2974,14 @@ def test_extended_axis(self):
...
-        assert_raises(IndexError, np.percentile, d, axis=4, q=25)
-        assert_raises(IndexError, np.percentile, d, axis=(0, 4), q=25)
+        assert_raises(np.AxisError, np.percentile, d, axis=-5, q=25)
+        assert_raises(np.AxisError, np.percentile, d, axis=(0, -5), q=25)
+        assert_raises(np.AxisError, np.percentile, d, axis=4, q=25)
+        assert_raises(np.AxisError, np.percentile, d, axis=(0, 4), q=25)
+        # each of these refers to the same axis twice
         assert_raises(ValueError, np.percentile, d, axis=(1, 1), q=25)
+        assert_raises(ValueError, np.percentile, d, axis=(-1, -1), q=25)
+        assert_raises(ValueError, np.percentile, d, axis=(3, -1), q=25)
```
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py]: These test the better duplicate-axis detection

File: numpy/lib/tests/test_function_base.py [hunk 3]
```diff
@@ -902,9 +902,9 @@ def test_specific_axes(self):
         # test maximal number of varargs
         assert_raises(TypeError, gradient, x, 1, 2, axis=1)
 
-        assert_raises(ValueError, gradient, x, axis=3)
-        assert_raises(ValueError, gradient, x, axis=-3)
-        assert_raises(TypeError, gradient, x, axis=[1,])
+        assert_raises(np.AxisError, gradient, x, axis=3)
```
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py]: These exceptions were found by temporarily removing the base classes from `AxisError`, so that tests for `ValueError` and `IndexError` would not catch them.

File: numpy/lib/function_base.py [hunk 4]
```diff
@@ -4764,6 +4748,8 @@ def delete(arr, obj, axis=None):
         else:
             return arr.copy(order=arrorder)
 
+    axis = normalize_axis_index(axis, ndim)
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: This guards against `IndexError`

File: numpy/core/tests/test_numeric.py [hunk 5]
```diff
@@ -2425,11 +2425,11 @@ def test_move_multiples(self):
...
     def test_errors(self):
         x = np.random.randn(1, 2, 3)
-        assert_raises_regex(np.AxisError, 'invalid axis .* `source`',
+        assert_raises_regex(np.AxisError, 'source.*out of bounds',
                             np.moveaxis, x, 3, 0)
-        assert_raises_regex(np.AxisError, 'invalid axis .* `source`',
+        assert_raises_regex(np.AxisError, 'source.*out of bounds',
                             np.moveaxis, x, -4, 0)
-        assert_raises_regex(np.AxisError, 'invalid axis .* `destination`',
+        assert_raises_regex(np.AxisError, 'destination.*out of bounds',
```
Comment (eric-wieser) [inline numpy/core/tests/test_numeric.py]: Error message comparison:

```
ValueError: invalid axis for this array in `destination` argument
AxisError: destination: axis 5 is out of bounds for array of dimension 3
```

File: numpy/lib/tests/test_function_base.py [hunk 6]
```diff
@@ -902,9 +902,9 @@ def test_specific_axes(self):
...
+        assert_raises(np.AxisError, gradient, x, axis=-3)
+        # assert_raises(TypeError, gradient, x, axis=[1,])
```
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py]: This used to error, and ideally it would still error for consistency with `ufunc.reduce`. But all the functions using `_ureduce` do not error in this situation, so we've already failed at consistency.

Perhaps we should add a deprecation warning (in a future commit) about passing lists of axes?

File: numpy/core/_internal.py [hunk 7]
```diff
@@ -631,4 +631,17 @@ class TooHardError(RuntimeError):
...
 class AxisError(ValueError, IndexError):
-    pass
+    def __init__(self, axis, ndim=None, msg_prefix=None):
+        # single-argument form just delegates to base class
+        if ndim is None and msg_prefix is None:
+            msg = axis
+
+        # do the string formatting here, to save work in the C code
+        else:
+            msg = "axis {} is out of bounds for array of dimension {}".format(
```
Comment (mhvk) [inline numpy/core/_internal.py]: Nitpicking, but I find the following much more readable
```
            msg = ("axis {} is out of bounds for array of dimension {}"
                   .format(axis, ndim))
```

File: numpy/core/numeric.py [hunk 8]
```diff
@@ -1544,17 +1542,59 @@ def rollaxis(a, axis, start=0):
     return a.transpose(axes)
 
 
-def _validate_axis(axis, ndim, argname):
+def normalize_axis_tuple(axis, ndim, argname=None, allow_duplicate=False):
+    """
+    Normalizes a tuple of axes, such that they are  valid positive indices
```
Comment (mhvk) [inline numpy/core/numeric.py]: Should have a single line as "title" (with a full stop at the end)

File: numpy/core/numeric.py [hunk 9]
```diff
@@ -1544,17 +1542,59 @@ def rollaxis(a, axis, start=0):
...
+    axis : int, tuple of int, or list of int
+        The un-normalized index or indices of the axis.
+    ndim : int
+        The number of dimensions of the array that `axis` should be normalized
+        against
+    argname : str, optional
+        A prefix to put before the error message, typically the name of the
+        argument
+    allow_duplicate : bool, optional
+        If True, the default, disallow an axis from being specified twice
```
Comment (mhvk) [inline numpy/core/numeric.py]: If `False`!

File: numpy/core/numeric.py [hunk 10]
```diff
@@ -1544,17 +1542,59 @@ def rollaxis(a, axis, start=0):
...
+    By default, disallows duplicates
+
+    Used internally by multi-axis-checking logic.
+
+    .. versionadded:: 1.13.0
+
+    Parameters
+    ----------
+    axis : int, tuple of int, or list of int
+        The un-normalized index or indices of the axis.
```
Comment (mhvk) [inline numpy/core/numeric.py]: The explanation should have a full stop at the end (here and below); I think this actually matters for sphinx rendering...

File: numpy/core/numeric.py [hunk 11]
```diff
@@ -1544,17 +1542,59 @@ def rollaxis(a, axis, start=0):
...
-    axis = [a + ndim if a < 0 else a for a in axis]
-    if not builtins.all(0 <= a < ndim for a in axis):
-        raise AxisError('invalid axis for this array in `%s` argument' %
-                         argname)
-    if len(set(axis)) != len(axis):
-        raise ValueError('repeated axis in `%s` argument' % argname)
+    axis = [normalize_axis_index(ax, ndim, argname) for ax in axis]
+    if not allow_duplicate and len(set(axis)) != len(axis):
+        if argname:
+            raise ValueError('repeated axis in `%s` argument' % argname)
```
Comment (mhvk) [inline numpy/core/numeric.py]: May as well change to new-style formatting.

**Review by mhvk — COMMENTED** [review]
This looks great! only rather minor comments.

File: numpy/core/_internal.py (hunk 7 above)
Comment (MSeifert04) [inline numpy/core/_internal.py]: or `f"axis {axis} is out of bounds for array of dimension {ndim}"` and force everyone to use Python 3.6. 😈

File: numpy/core/numeric.py [hunk 12]
```diff
@@ -1544,17 +1542,59 @@ def rollaxis(a, axis, start=0):
     return a.transpose(axes)
 
 
-def _validate_axis(axis, ndim, argname):
+def normalize_axis_tuple(axis, ndim, argname=None, allow_duplicate=False):
```
Comment (mhvk) [inline numpy/core/numeric.py]: I slightly wonder -- mostly because of #8819 -- whether it would be useful to have this in C...

Comment (eric-wieser) [inline numpy/core/numeric.py]: >  whether it would be useful to have this in C

Yes, I think it would - and I think we even already have it for `ufunc.reduce`. I think this is a good stepping stone to implementing that, since it pulls all the python codepaths through one point for parsing axes - that point can now be moved to C later.

File: numpy/core/numeric.py (hunk 11 above)
Comment (eric-wieser) [inline numpy/core/numeric.py]: Sensible

File: numpy/core/numeric.py (hunk 9 above)
Comment (eric-wieser) [inline numpy/core/numeric.py]: Oops!

File: numpy/core/_internal.py [hunk 13]
```diff
@@ -631,4 +631,17 @@ class TooHardError(RuntimeError):
     pass
 
 class AxisError(ValueError, IndexError):
-    pass
+    """ Axis supplied was invalid. """
```
Comment (eric-wieser) [inline numpy/core/_internal.py]: Now with a docstring, which Ipython shows when you type an exception name. This is in a similar style to the builtin ones.

**Review by eric-wieser — COMMENTED** [review]
Ok, all fixed up

File: numpy/core/_internal.py (hunk 13 above)
Comment (charris) [inline numpy/core/_internal.py]: This should be in `numpy/_globals.py`.

Comment (charris) [inline numpy/core/_internal.py]: And imported into `numpy/__init__.py`.

Comment (eric-wieser) [inline numpy/core/_internal.py]: I put it here in #8584 because it needs to be imported from C code, and that's what `_internals` seems to be used for. Is it safe to load `_globals.py` from C?

Comment (charris) [inline numpy/core/_internal.py]: You can import from practically anything. We also import from multiarray itself into C code. I'd import this from '"numpy"', The `_globals` is for singletons that may also be used by python code.

Comment (charris) [inline numpy/core/_internal.py]: Python code should not be importing from `_internal.py`.

Comment (eric-wieser) [inline numpy/core/_internal.py]: Should `TooHardError` move there too?

File: numpy/core/src/multiarray/common.h (hunk 1 above)
Comment (eric-wieser) [inline numpy/core/src/multiarray/common.h]: Wait I should just use `PyObject_CallFunction` here

File: numpy/core/_internal.py (hunk 13 above)
Comment (eric-wieser) [inline numpy/core/_internal.py]: right now, `numeric.py` has `from ._internal import TooHardError, AxisError`

`core/__init__.py` also contains `from . import _internal  # for freeze programs`, for some reason

Comment (charris) [inline numpy/core/_internal.py]: Probably, but it can be done in a separate PR as the imports also need to be changed. Maybe `_globals.py` should be renamed `_global_singletons` or `_global_exceptions` ;) There are other errors scattered about that maybe we should move at some point.

Comment (charris) [inline numpy/core/_internal.py]: Or maybe `_numpy_exceptions`

Comment (eric-wieser) [inline numpy/core/_internal.py]: Do you want me to do both in a separate PR? `AxisError` being in that file is not new to this PR.

File: numpy/core/src/multiarray/common.h [hunk 14]
```diff
@@ -158,9 +162,15 @@ check_and_adjust_axis(int *axis, int ndim)
             }
         }
 
-        PyErr_Format(AxisError_cls,
-                     "axis %d is out of bounds for array of dimension %d",
-                     *axis, ndim);
+        /* Invoke the AxisError constructor */
+        exc = PyObject_CallFunction(AxisError_cls, "iiO",
+                                    *axis, ndim, msg_prefix);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/common.h]: This was the way I should have written this the first time...

**Comment by eric-wieser** [comment]
Ok, this is rebased over merge conflicts, using more appropriate bits of the python API to construct the exception, and passes.

As @charris points out, this exception might be in the wrong place. But thats:
 * Not added by this commit
 * Not the only exception with this problem

So should probably be moved to another issue/PR?

File: numpy/core/numeric.py [hunk 15]
```diff
@@ -1544,17 +1542,59 @@ def rollaxis(a, axis, start=0):
...
         axis = [operator.index(axis)]
     except TypeError:
         axis = list(axis)
-    axis = [a + ndim if a < 0 else a for a in axis]
-    if not builtins.all(0 <= a < ndim for a in axis):
-        raise AxisError('invalid axis for this array in `%s` argument' %
-                         argname)
-    if len(set(axis)) != len(axis):
-        raise ValueError('repeated axis in `%s` argument' % argname)
+    axis = [normalize_axis_index(ax, ndim, argname) for ax in axis]
```
Comment (mhvk) [inline numpy/core/numeric.py]: In the docstring, you write that you'll return a tuple, but you actually construct a list. Just replace `[...] with `tuple(...)` (also in the try/except above, for completeness.)

File: numpy/core/src/multiarray/common.h [hunk 16]
```diff
@@ -169,6 +179,11 @@ check_and_adjust_axis(int *axis, int ndim)
     }
     return 0;
 }
+static NPY_INLINE int
+check_and_adjust_axis(int *axis, int ndim)
+{
+    return check_and_adjust_axis_msg(axis, ndim, Py_None);
```
Comment (mhvk) [inline numpy/core/src/multiarray/common.h]: I think you have to `Py_INCREF(Py_None)` before passing it on. Since you then would have to `Py_DECREF` it as well, I suggest passing in `NULL` here, and doing the translation to `Py_None` above.

File: numpy/core/src/multiarray/multiarraymodule.c [hunk 17]
```diff
@@ -4106,16 +4106,16 @@ array_may_share_memory(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *
 static PyObject *
 normalize_axis_index(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwds)
 {
-    static char *kwlist[] = {"axis", "ndim", NULL};
+    static char *kwlist[] = {"axis", "ndim", "msg_prefix", NULL};
     int axis;
     int ndim;
+    PyObject *msg_prefix = Py_None;
```
Comment (mhvk) [inline numpy/core/src/multiarray/multiarraymodule.c]: Here, again, I think you cannot just use `Py_None`. Also solve by initlializing to `NULL`?

**Review by mhvk — CHANGES_REQUESTED** [review]

File: numpy/core/src/multiarray/common.h (hunk 16 above)
Comment (MSeifert04) [inline numpy/core/src/multiarray/common.h]: `Py_None` is garantueed to exist and as long as it doesn't escape to the user and it isn't decref'd everything should be fine. `PyObject_CallFunction` increments it's reference count for the exception so what the user gets is a new reference.

Using `NULL` seems cleaner though.

Comment (eric-wieser) [inline numpy/core/src/multiarray/common.h]: I don't need to here, because `check_and_adjust_axis_msg` takes a borrowed reference. It doesn't actually need a refcount at all until it makes it to the python code, and `PyObject_CallFunction()` increments the refcount for me, AFAICT.

File: numpy/core/src/multiarray/multiarraymodule.c (hunk 17 above)
Comment (eric-wieser) [inline numpy/core/src/multiarray/multiarraymodule.c]: This is definitely correct. `PyArg_ParseTuple` does not `INCREF` `O` arguments, so this line is exactly equivalent to passing `None` from python.

The question remains if I should `INCREF` before calling `check_and_adjust_axis_msg`, but let's discuss that above.

File: numpy/core/numeric.py (hunk 15 above)
Comment (eric-wieser) [inline numpy/core/numeric.py]: Yep, that's a bug.

File: numpy/core/src/multiarray/common.h (hunk 16 above)
Comment (eric-wieser) [inline numpy/core/src/multiarray/common.h]: > Using NULL seems cleaner though.

Why? We already need to handle `Py_None` being passed in from user code. Furthermore, when we get `Py_None` from usercode, it's a borrowed reference - so like here, it doesn't need an incref unless it escapes to python code.

Comment (MSeifert04) [inline numpy/core/src/multiarray/common.h]: @eric-wieser How would it cause segfaults? You could just initialize it as `NULL` and when you call the "exception" you do `PyObject_CallFunction(AxisError_cls, "iiO", *axis, ndim, msg_prefix?msg_prefix:Py_None);`

Comment (eric-wieser) [inline numpy/core/src/multiarray/common.h]: Wait, why on earth is `msg_prefix?msg_prefix:Py_None` better than just passing `Py_None` in the first place? The refcounting is identical

Comment (MSeifert04) [inline numpy/core/src/multiarray/common.h]: That wasn't meant as the "cleaner" alternative. I just wondered how that could segfault.

Comment (eric-wieser) [inline numpy/core/src/multiarray/common.h]: Segfaults came from any other kind of refcounting. I wasn't aiming to be passing null.

Comment (MSeifert04) [inline numpy/core/src/multiarray/common.h]: Yeah, that's the reason why it's generally not a good idea to pass borrowed references around. But I didn't check the numpy source code, maybe it's common here. In that case please ignore my comments.

Comment (eric-wieser) [inline numpy/core/src/multiarray/common.h]: I guess my argument would be that if I manually inlined the contents of `check_and_adjust_axis_msg` into `normalize_axis_index`, you would not expect me to add any additional `Py_INCREF`s

I'm not sure what you mean there - can you elaborate on exactly why it is a bad idea?

File: numpy/core/numeric.py (hunk 15 above)
Comment (eric-wieser) [inline numpy/core/numeric.py]: Done - fixed to return a tuple. This is actually pretty important, given that things like `sum` require `axis` to be a tuple, so would previously not work for any kind of normalize-and-forward usage.

Thanks for the catch!

File: numpy/core/src/multiarray/common.h (hunk 16 above)
Comment (MSeifert04) [inline numpy/core/src/multiarray/common.h]: ok, this goes a bit off-topic. I [still think](https://github.com/numpy/numpy/pull/8843#discussion_r108532657) the code is correct wrt reference counting.

Comment (eric-wieser) [inline numpy/core/src/multiarray/common.h]: Some confusion added here by my first comment being written before yours appeared (and not directed at you), I think :).

**Comment by mhvk** [comment]
@eric-wieser - I think you convinced me that your approach works ref-counting wise, but I'm still not quite convinced it is the handiest. My suggestion would be to simply pass string around in the C facing routines, i.e., make `check_and_adjust_axis_msg` more easily useful also from within `C` and create a corresponding object only when setting the `AxisError`. I realise this implies possibly getting the string from a `Python` object, and then recreating the object for the error message, but think this is OK.

Anyway, it is not a big deal.

**Review by mhvk — APPROVED** [review]

**Comment by eric-wieser** [comment]
> would be to simply pass string around in the C facing routines, i.e., make check_and_adjust_axis_msg more easily useful also from within C 

Seems reasonable. Although I didn't find myself needing any extra info in the C code, since everything there seems to be working with a single axis argument. So I think we do that if we need it, and otherwise don't bother.

> and create a corresponding object only when setting the AxisError

Alternatively, the C code could just have a static pre-allocated error-message string

**Comment by juliantaylor** [comment]
looks good to me

**Comment by mhvk** [comment]
> Seems reasonable. Although I didn't find myself needing any extra info in the C code, since everything there seems to be working with a single axis argument. So I think we do that if we need it, and otherwise don't bother.

The only issue with this is that if you leave this in place, if anyone in the C code is going to use `check_and_adjust_axis_msg`, i.e., they want to make the error that might get raised more descriptive, they'll set up a `PyObject` and then it just becomes harder to undo. So, personally, I'd prefer to get it right now (but I realise I may be thinking too far ahead, so if you don't agree, that's OK, and we can just merge as is).

**Comment by eric-wieser** [comment]
I think that if anyone doing that themselves finds they do want to be able to pass a c string, they'll be in a more capable decision to decide if they should just do it in place. We already create python strings in lots of places for error messages.

Besides, to support this, I think we'd need a `..._msg_or_cmsg(axis, ndim, PyObject *msg, char* cmsg)` which could support both. This seems like a messy addition when it is not even needed yet.

**Comment by mhvk** [comment]
OK, fair enough. Since both @juliantaylor and I approved this, let's just go ahead and merge.

**Comment by eric-wieser** [comment]
Thanks @mhvk!
